File: python/send_email.py
#!/usr/bin/env python3
"""
Gmail Hotkey Sender - Python Implementation

A CLI tool for sending Gmail emails via keyboard shortcuts.
This is a starter implementation that requires Gmail API setup.
"""

import argparse
import json
import logging
import os
import sys
from pathlib import Path
from typing import Dict, List, Optional

# TODO: Add these imports when implementing Gmail API
# from google.auth.transport.requests import Request
# from google.oauth2.credentials import Credentials
# from google_auth_oauthlib.flow import InstalledAppFlow
# from googleapiclient.discovery import build
# from googleapiclient.errors import HttpError

# Configure logging
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(levelname)s - %(message)s'
)
logger = logging.getLogger(__name__)

# Gmail API scopes
SCOPES = ['https://www.googleapis.com/auth/gmail.send']

class GmailSender:
    """Gmail email sender using Gmail API."""
    
    def __init__(self, config_path: Optional[str] = None):
        """Initialize the Gmail sender.
        
        Args:
            config_path: Path to configuration file
        """
        self.config_path = config_path or self._get_default_config_path()
        self.config = self._load_config()
        self.service = None
        
    def _get_default_config_path(self) -> str:
        """Get the default configuration file path."""
        config_dir = Path.home() / '.gmail-hotkey-sender'
        config_dir.mkdir(exist_ok=True)
        return str(config_dir / 'config.json')
    
    def _load_config(self) -> Dict:
        """Load configuration from file."""
        if os.path.exists(self.config_path):
            try:
                with open(self.config_path, 'r') as f:
                    return json.load(f)
            except json.JSONDecodeError:
                logger.warning(f"Invalid JSON in config file: {self.config_path}")
        
        # Default configuration
        default_config = {
            'credentials_file': 'credentials.json',
            'token_file': 'token.json',
            'default_subject': '',
            'default_body': '',
            'rate_limit': 100,  # emails per day
            'log_sent_emails': True
        }
        
        # Save default config
        self._save_config(default_config)
        return default_config
    
    def _save_config(self, config: Dict) -> None:
        """Save configuration to file."""
        try:
            with open(self.config_path, 'w') as f:
                json.dump(config, f, indent=2)
        except Exception as e:
            logger.error(f"Failed to save config: {e}")
    
    def authenticate(self) -> bool:
        """Authenticate with Gmail API.
        
        Returns:
            True if authentication successful, False otherwise
        """
        # TODO: Implement Gmail API authentication
        logger.info("Gmail API authentication not yet implemented")
        logger.info("Please implement OAuth2 flow with google-auth-oauthlib")
        return False
    
    def send_email(self, to: str, subject: str, body: str, 
                   cc: Optional[List[str]] = None, 
                   bcc: Optional[List[str]] = None,
                   html_body: Optional[str] = None) -> bool:
        """Send an email via Gmail API.
        
        Args:
            to: Recipient email address(es)
            subject: Email subject
            body: Plain text email body
            cc: CC recipients
            bcc: BCC recipients
            html_body: HTML email body (optional)
            
        Returns:
            True if email sent successfully, False otherwise
        """
        if not self.service:
            if not self.authenticate():
                logger.error("Failed to authenticate with Gmail API")
                return False
        
        try:
            # TODO: Implement actual email sending with Gmail API
            logger.info(f"Would send email to: {to}")
            logger.info(f"Subject: {subject}")
            logger.info(f"Body: {body[:100]}...")
            
            # Log the email if enabled
            if self.config.get('log_sent_emails', True):
                self._log_email(to, subject, body)
            
            return True
            
        except Exception as e:
            logger.error(f"Failed to send email: {e}")
            return False
    
    def _log_email(self, to: str, subject: str, body: str) -> None:
        """Log sent email details."""
        log_dir = Path.home() / '.gmail-hotkey-sender' / 'logs'
        log_dir.mkdir(exist_ok=True)
        
        log_file = log_dir / 'sent_emails.log'
        timestamp = logging.Formatter().formatTime(logging.LogRecord(
            'gmail_sender', logging.INFO, '', 0, '', (), None
        ))
        
        with open(log_file, 'a') as f:
            f.write(f"{timestamp} | To: {to} | Subject: {subject}\n")
    
    def create_draft(self, to: str, subject: str, body: str) -> bool:
        """Create a draft email in Gmail.
        
        Args:
            to: Recipient email address(es)
            subject: Email subject
            body: Email body
            
        Returns:
            True if draft created successfully, False otherwise
        """
        # TODO: Implement draft creation
        logger.info(f"Would create draft to: {to}")
        return True

def main():
    """Main CLI entry point."""
    parser = argparse.ArgumentParser(
        description="Send Gmail emails via CLI with hotkey support"
    )
    
    parser.add_argument(
        '--to', '-t',
        required=True,
        help='Recipient email address(es)'
    )
    
    parser.add_argument(
        '--subject', '-s',
        default='',
        help='Email subject'
    )
    
    parser.add_argument(
        '--body', '-b',
        default='',
        help='Email body'
    )
    
    parser.add_argument(
        '--cc',
        nargs='+',
        help='CC recipient(s)'
    )
    
    parser.add_argument(
        '--bcc',
        nargs='+',
        help='BCC recipient(s)'
    )
    
    parser.add_argument(
        '--html',
        help='HTML email body (optional)'
    )
    
    parser.add_argument(
        '--draft',
        action='store_true',
        help='Create draft instead of sending'
    )
    
    parser.add_argument(
        '--config',
        help='Path to configuration file'
    )
    
    parser.add_argument(
        '--verbose', '-v',
        action='store_true',
        help='Enable verbose logging'
    )
    
    args = parser.parse_args()
    
    if args.verbose:
        logging.getLogger().setLevel(logging.DEBUG)
    
    # Initialize Gmail sender
    sender = GmailSender(args.config)
    
    # Send or create draft
    if args.draft:
        success = sender.create_draft(args.to, args.subject, args.body)
    else:
        success = sender.send_email(
            to=args.to,
            subject=args.subject,
            body=args.body,
            cc=args.cc,
            bcc=args.bcc,
            html_body=args.html
        )
    
    if success:
        logger.info("Operation completed successfully")
        sys.exit(0)
    else:
        logger.error("Operation failed")
        sys.exit(1)

if __name__ == '__main__':
    main()

File: nodejs/send_email.js
#!/usr/bin/env node
/**
 * Gmail Hotkey Sender - Node.js Implementation
 * 
 * A CLI tool for sending Gmail emails via keyboard shortcuts.
 * This is a starter implementation that requires Gmail API setup.
 */

const fs = require('fs');
const path = require('path');
const os = require('os');

// TODO: Add these imports when implementing Gmail API
// const { google } = require('googleapis');
// const { OAuth2Client } = require('google-auth-library');

/**
 * Gmail email sender using Gmail API
 */
class GmailSender {
    constructor(configPath = null) {
        this.configPath = configPath || this.getDefaultConfigPath();
        this.config = this.loadConfig();
        this.auth = null;
        this.gmail = null;
    }

    /**
     * Get the default configuration file path
     */
    getDefaultConfigPath() {
        const configDir = path.join(os.homedir(), '.gmail-hotkey-sender');
        if (!fs.existsSync(configDir)) {
            fs.mkdirSync(configDir, { recursive: true });
        }
        return path.join(configDir, 'config.json');
    }

    /**
     * Load configuration from file
     */
    loadConfig() {
        if (fs.existsSync(this.configPath)) {
            try {
                const configData = fs.readFileSync(this.configPath, 'utf8');
                return JSON.parse(configData);
            } catch (error) {
                console.warn(`Invalid JSON in config file: ${this.configPath}`);
            }
        }

        // Default configuration
        const defaultConfig = {
            credentialsFile: 'credentials.json',
            tokenFile: 'token.json',
            defaultSubject: '',
            defaultBody: '',
            rateLimit: 100, // emails per day
            logSentEmails: true
        };

        // Save default config
        this.saveConfig(defaultConfig);
        return defaultConfig;
    }

    /**
     * Save configuration to file
     */
    saveConfig(config) {
        try {
            fs.writeFileSync(this.configPath, JSON.stringify(config, null, 2));
        } catch (error) {
            console.error(`Failed to save config: ${error.message}`);
        }
    }

    /**
     * Authenticate with Gmail API
     */
    async authenticate() {
        // TODO: Implement Gmail API authentication
        console.log('Gmail API authentication not yet implemented');
        console.log('Please implement OAuth2 flow with googleapis');
        return false;
    }

    /**
     * Send an email via Gmail API
     */
    async sendEmail(to, subject, body, cc = null, bcc = null, htmlBody = null) {
        if (!this.gmail) {
            if (!(await this.authenticate())) {
                console.error('Failed to authenticate with Gmail API');
                return false;
            }
        }

        try {
            // TODO: Implement actual email sending with Gmail API
            console.log(`Would send email to: ${to}`);
            console.log(`Subject: ${subject}`);
            console.log(`Body: ${body.substring(0, 100)}...`);

            // Log the email if enabled
            if (this.config.logSentEmails) {
                this.logEmail(to, subject, body);
            }

            return true;
        } catch (error) {
            console.error(`Failed to send email: ${error.message}`);
            return false;
        }
    }

    /**
     * Log sent email details
     */
    logEmail(to, subject, body) {
        const logDir = path.join(os.homedir(), '.gmail-hotkey-sender', 'logs');
        if (!fs.existsSync(logDir)) {
            fs.mkdirSync(logDir, { recursive: true });
        }

        const logFile = path.join(logDir, 'sent_emails.log');
        const timestamp = new Date().toISOString();
        const logEntry = `${timestamp} | To: ${to} | Subject: ${subject}\n`;

        fs.appendFileSync(logFile, logEntry);
    }

    /**
     * Create a draft email in Gmail
     */
    async createDraft(to, subject, body) {
        // TODO: Implement draft creation
        console.log(`Would create draft to: ${to}`);
        return true;
    }
}

/**
 * Parse command line arguments
 */
function parseArgs() {
    const args = process.argv.slice(2);
    const parsed = {
        to: null,
        subject: '',
        body: '',
        cc: null,
        bcc: null,
        html: null,
        draft: false,
        config: null,
        verbose: false
    };

    for (let i = 0; i < args.length; i++) {
        const arg = args[i];
        
        switch (arg) {
            case '--to':
            case '-t':
                parsed.to = args[++i];
                break;
            case '--subject':
            case '-s':
                parsed.subject = args[++i];
                break;
            case '--body':
            case '-b':
                parsed.body = args[++i];
                break;
            case '--cc':
                parsed.cc = args[++i].split(',');
                break;
            case '--bcc':
                parsed.bcc = args[++i].split(',');
                break;
            case '--html':
                parsed.html = args[++i];
                break;
            case '--draft':
                parsed.draft = true;
                break;
            case '--config':
                parsed.config = args[++i];
                break;
            case '--verbose':
            case '-v':
                parsed.verbose = true;
                break;
            case '--help':
            case '-h':
                showHelp();
                process.exit(0);
                break;
        }
    }

    return parsed;
}

/**
 * Show help information
 */
function showHelp() {
    console.log(`
Gmail Hotkey Sender - Node.js Implementation

Usage: node send_email.js [options]

Options:
  --to, -t <email>        Recipient email address(es) (required)
  --subject, -s <text>    Email subject
  --body, -b <text>       Email body
  --cc <emails>           CC recipient(s) (comma-separated)
  --bcc <emails>          BCC recipient(s) (comma-separated)
  --html <html>           HTML email body (optional)
  --draft                 Create draft instead of sending
  --config <path>         Path to configuration file
  --verbose, -v           Enable verbose logging
  --help, -h              Show this help message

Examples:
  node send_email.js --to alice@example.com --subject "Test" --body "Hello"
  node send_email.js -t bob@example.com -s "Meeting" -b "Reminder" --draft
`);
}

/**
 * Main function
 */
async function main() {
    const args = parseArgs();

    if (args.verbose) {
        // Enable verbose logging
        console.log('Verbose mode enabled');
    }

    if (!args.to) {
        console.error('Error: --to argument is required');
        showHelp();
        process.exit(1);
    }

    // Initialize Gmail sender
    const sender = new GmailSender(args.config);

    try {
        // Send or create draft
        let success;
        if (args.draft) {
            success = await sender.createDraft(args.to, args.subject, args.body);
        } else {
            success = await sender.sendEmail(
                args.to,
                args.subject,
                args.body,
                args.cc,
                args.bcc,
                args.html
            );
        }

        if (success) {
            console.log('Operation completed successfully');
            process.exit(0);
        } else {
            console.error('Operation failed');
            process.exit(1);
        }
    } catch (error) {
        console.error(`Unexpected error: ${error.message}`);
        process.exit(1);
    }
}

// Run the main function if this file is executed directly
if (require.main === module) {
    main();
}

module.exports = { GmailSender };
